src/mainboard/clevo/adl-p/ramstage.c

```diff
@@ -65,9 +108,54 @@ void mainboard_silicon_init_params(FSP_S_CONFIG *params)
 	params->CpuPcieRpAcsEnabled[0] = 1;
 }
 
+static void set_fan_curve(void)
+{
+	int i;
+	uint8_t selection = FAN_CURVE_OPTION_DEFAULT;
+	struct smfi_cmd_set_fan_curve cmd;
+
+#if CONFIG(DRIVERS_EFI_VARIABLE_STORE)
+	struct region_device rdev;
+	enum cb_err ret;
+	uint32_t size;
+
+	const EFI_GUID dasharo_system_features_guid = {
+		0xd15b327e, 0xff2d, 0x4fc1, { 0xab, 0xf6, 0xc1, 0x2b, 0xd0, 0x8c, 0x13, 0x59 }
+	};
+
+	if (smmstore_lookup_region(&rdev))
+		goto efi_err;
+
+	size = sizeof(selection);
+	ret = efi_fv_get_option(&rdev, &dasharo_system_features_guid, "FanCurveOption", &selection, &size);
+	if (ret != CB_SUCCESS)
+		printk(BIOS_DEBUG, "fan: failed to read curve selection from EFI vars, using board default\n");
+efi_err:
+#endif
+
+	switch (selection) {
+	case FAN_CURVE_OPTION_SILENT:
+		cmd.curve = fan_curve_silent;
```

Comment:
does such assignment work as expected? For simple structures it probably would, but to be sure memcpy will be safest. @krystian-hebel can you confirm?

Reply:
In this case it doesn't make a difference, basic assignment is safe to use. `memcpy` helps in two cases:
- an incomplete type for which size of copied data isn't available, e.g. when structure ends with unbounded array (like `uint8_t buf[];`) or when more than one element of array is to be copied, or when compiler doesn't have full definition of type,
- struct has unaligned fields on architectures that can't access such data directly (but this usually is mitigated by compiler anyway).

Neither of those would help if the structure has pointers to data that would also have to be copied.